File: app/models/transaction.py

# app/models/transaction.py
from sqlalchemy import String, DateTime, Enum as SAEnum, ForeignKey, Text
from sqlalchemy.orm import Mapped, mapped_column, relationship
from enum import Enum
from datetime import datetime
from app.database import Base
from app.core.datetime_utils import now_sp

class TipoMovimentacao(str, Enum):
    EMPRESTIMO = "empréstimo"
    DEVOLUCAO = "devolução"
    TRANSFERENCIA = "transferência"
    MANUTENCAO = "manutenção"
    BAIXA = "baixa"
    CADASTRO = "cadastro"

class StatusSolicitacao(str, Enum):
    PENDENTE = "Pendente"
    APROVADA = "Aprovada"
    ENTREGUE = "Entregue"  # Novo - confirmação de recebimento
    REJEITADA = "Rejeitada"
    CANCELADA = "Cancelada"

class Movimentacao(Base):
    __tablename__ = "movimentacoes"

    id: Mapped[int] = mapped_column(primary_key=True, index=True)
    asset_id: Mapped[int] = mapped_column(ForeignKey("assets.id"), nullable=False)
    data: Mapped[datetime] = mapped_column(DateTime, default=now_sp)
    tipo: Mapped[TipoMovimentacao] = mapped_column(SAEnum(TipoMovimentacao), nullable=False)
    
    # Origem e Destino (Users)
    de_user_id: Mapped[int | None] = mapped_column(ForeignKey("users.id"), nullable=True)
    para_user_id: Mapped[int | None] = mapped_column(ForeignKey("users.id"), nullable=True)
    
    # Origem e Destino (Departamentos - opcional, dependendo da lógica)
    de_departamento_id: Mapped[int | None] = mapped_column(ForeignKey("departamentos.id"), nullable=True)
    para_departamento_id: Mapped[int | None] = mapped_column(ForeignKey("departamentos.id"), nullable=True)
    
    observacao: Mapped[str | None] = mapped_column(Text, nullable=True)

    # Relacionamentos
    asset = relationship("Asset", back_populates="movimentacoes")
    de_user = relationship("User", foreign_keys=[de_user_id], back_populates="movimentacoes_origem")
    para_user = relationship("User", foreign_keys=[para_user_id], back_populates="movimentacoes_destino")


class Solicitacao(Base):
    __tablename__ = "solicitacoes"

    id: Mapped[int] = mapped_column(primary_key=True, index=True)
    solicitante_id: Mapped[int] = mapped_column(ForeignKey("users.id"), nullable=False)
    asset_id: Mapped[int | None] = mapped_column(ForeignKey("assets.id"), nullable=True) # Pode solicitar um tipo, não necessariamente um asset específico, mas por enquanto linkamos direto se souber
    
    data_solicitacao: Mapped[datetime] = mapped_column(DateTime, default=now_sp)
    motivo: Mapped[str] = mapped_column(Text, nullable=False)
    status: Mapped[StatusSolicitacao] = mapped_column(SAEnum(StatusSolicitacao), default=StatusSolicitacao.PENDENTE)
    
    # Aprovação
    aprovador_id: Mapped[int | None] = mapped_column(ForeignKey("users.id"), nullable=True)
    data_aprovacao: Mapped[datetime | None] = mapped_column(DateTime, nullable=True)
    data_prevista_devolucao: Mapped[datetime | None] = mapped_column(DateTime, nullable=True)
    
    # Confirmação de Entrega (via QR Code ou manual)
    data_entrega: Mapped[datetime | None] = mapped_column(DateTime, nullable=True)
    confirmado_por_id: Mapped[int | None] = mapped_column(ForeignKey("users.id"), nullable=True)  # Admin/Gerente que confirmou
    confirmado_via_qr: Mapped[bool | None] = mapped_column(default=False)  # Se foi via QR do usuário
    observacao_entrega: Mapped[str | None] = mapped_column(Text, nullable=True)

    # Relacionamentos
    solicitante = relationship("User", foreign_keys=[solicitante_id], back_populates="solicitacoes")
    aprovador = relationship("User", foreign_keys=[aprovador_id], back_populates="aprovacoes")
    confirmador = relationship("User", foreign_keys=[confirmado_por_id])
    asset = relationship("Asset", back_populates="solicitacoes")

File: backend/internal/models/transaction.go
package models

import (
	"time"
)

type TipoMovimentacao string

const (
	TipoMovimentacaoEmprestimo    TipoMovimentacao = "empréstimo"
	TipoMovimentacaoDevolucao     TipoMovimentacao = "devolução"
	TipoMovimentacaoTransferencia TipoMovimentacao = "transferência"
	TipoMovimentacaoManutencao    TipoMovimentacao = "manutenção"
	TipoMovimentacaoBaixa         TipoMovimentacao = "baixa"
	TipoMovimentacaoCadastro      TipoMovimentacao = "cadastro"
)

type StatusSolicitacao string

const (
	StatusSolicitacaoPendente  StatusSolicitacao = "Pendente"
	StatusSolicitacaoAprovada  StatusSolicitacao = "Aprovada"
	StatusSolicitacaoEntregue  StatusSolicitacao = "Entregue"
	StatusSolicitacaoRejeitada StatusSolicitacao = "Rejeitada"
	StatusSolicitacaoCancelada StatusSolicitacao = "Cancelada"
)

type Movimentacao struct {
	ID                 uint             `gorm:"primaryKey" json:"id"`
	AssetID            uint             `gorm:"column:asset_id;not null" json:"asset_id"`
	Data               time.Time        `gorm:"column:data;autoCreateTime" json:"data"`
	Tipo               TipoMovimentacao `gorm:"type:varchar(30);not null" json:"tipo"`
	DeUserID           *uint            `gorm:"column:de_user_id" json:"de_user_id"`
	ParaUserID         *uint            `gorm:"column:para_user_id" json:"para_user_id"`
	DeDepartamentoID   *uint            `gorm:"column:de_departamento_id" json:"de_departamento_id"`
	ParaDepartamentoID *uint            `gorm:"column:para_departamento_id" json:"para_departamento_id"`
	Observacao         *string          `gorm:"type:text" json:"observacao"`

	Asset    *Asset `gorm:"foreignKey:AssetID" json:"asset,omitempty"`
	DeUser   *User  `gorm:"foreignKey:DeUserID" json:"de_user,omitempty"`
	ParaUser *User  `gorm:"foreignKey:ParaUserID" json:"para_user,omitempty"`
}

func (Movimentacao) TableName() string { return "movimentacoes" }

type Solicitacao struct {
	ID                    uint              `gorm:"primaryKey" json:"id"`
	SolicitanteID         *uint             `gorm:"column:solicitante_id" json:"solicitante_id"`
	AssetID               *uint             `gorm:"column:asset_id" json:"asset_id"`
	DataSolicitacao       time.Time         `gorm:"column:data_solicitacao;autoCreateTime" json:"data_solicitacao"`
	Motivo                string            `gorm:"type:text;not null" json:"motivo"`
	Status                StatusSolicitacao `gorm:"type:varchar(30);default:'Pendente'" json:"status"`
	AprovadorID           *uint             `gorm:"column:aprovador_id" json:"aprovador_id"`
	DataAprovacao         *time.Time        `gorm:"column:data_aprovacao" json:"data_aprovacao"`
	DataPrevistaDevolucao *time.Time        `gorm:"column:data_prevista_devolucao" json:"data_prevista_devolucao"`
	DataEntrega           *time.Time        `gorm:"column:data_entrega" json:"data_entrega"`
	ConfirmadoPorID       *uint             `gorm:"column:confirmado_por_id" json:"confirmado_por_id"`
	ConfirmadoViaQR       *bool             `gorm:"column:confirmado_via_qr;default:false" json:"confirmado_via_qr"`
	ObservacaoEntrega     *string           `gorm:"column:observacao_entrega;type:text" json:"observacao_entrega"`

	Solicitante *User  `gorm:"foreignKey:SolicitanteID" json:"solicitante,omitempty"`
	Aprovador   *User  `gorm:"foreignKey:AprovadorID" json:"aprovador,omitempty"`
	Confirmador *User  `gorm:"foreignKey:ConfirmadoPorID" json:"confirmador,omitempty"`
	Asset       *Asset `gorm:"foreignKey:AssetID" json:"asset,omitempty"`
}

func (Solicitacao) TableName() string { return "solicitacoes" }
